### PR TITLE
Copy package files from the package cache instead of ingesting them into the build cache

### PR DESCRIPTION
**Pull request by dfederm**
A non-trivial part of builds is simply copying content from packages and there's no need to duplicate that content into the build cache when the package cache should already contain the files. As packages are expected to be immutable, this should be a safe operation.

In my test repo with 100% cache hits, this saves ~50% of the build time (pulling everything from blob storage cache) and cuts the remote storage by about ~50% as well. Note that this test was on residential internet, so that 50% build time savings may not be completely representative if the CI build agents have a super fast connection to the storage, but the storage cost savings should be.

This change does a couple things:
* When processing a node in `HandleProjectFinished` ("cache add")
  * While enumerating the inputs, keep track of the ones under the nuget package root, hashing as necessary. Note that these files are hashed anyway when constructing either the weak or strong fingerprints, so this is not any additional hashing work.
  * While creating the `NodeBuildResult`, create a collection called `PackageFilesToCopy` which is the set of outputs whose hash matches a package input. The `PackageFilesToCopy` collection's keys are the enlistment-relative output paths and the values are the package root relative paths of the source file. This collection is added as part of the `NodeBuildResult`.
  * While adding content files to the cache, avoid adding any files listed in `PackageFilesToCopy`
* When fetching a node from the cache, specifically the content ("cache get")
  * If an output file is listed in `PackageFilesToCopy`, avoid asking the cache for it in favor of simply copying it from the package root. Note that CoW is supported for this operation.

### CODE REVIEW

File: src/Common/Caching/CacheClient.cs [hunk 1]
```diff
@@ -419,6 +466,14 @@ protected void CreateParentDirectory(string filePath)
         }
 
         return (pathSet, nodeBuildResult);
+
+        static bool AreOnSameDrive(string path1, string path2)
```
Comment (johnterickson) [inline src/Common/Caching/CacheClient.cs:470]: @erikmav what is the canonical way to do this?

File: src/Common/NodeBuildResult.cs [hunk 2]
```diff
@@ -18,12 +18,14 @@ public sealed class NodeBuildResult
     [JsonConstructor]
     public NodeBuildResult(
         SortedDictionary<string, ContentHash> outputs,
+        SortedDictionary<string, string> packageFilesToCopy,
```
Comment (johnterickson) [inline src/Common/NodeBuildResult.cs:21]: Bump CurrentVersion

File: src/Common/Caching/CacheClient.cs [hunk 3]
```diff
@@ -394,10 +407,44 @@ protected void CreateParentDirectory(string filePath)
...
+                string destinationAbsolutePath = Path.Combine(RepoRoot, kvp.Key);
+                if (nodeBuildResult.PackageFilesToCopy.TryGetValue(kvp.Key, out string? packageFile))
+                {
+                    tasks.Add(Task.Run(
+                        () =>
+                        {
+                            string sourceAbsolutePath = Path.Combine(_nugetPackageRoot, packageFile);
+                            CreateParentDirectory(destinationAbsolutePath);
+
+                            if (_canCloneInNugetCachePath && AreOnSameDrive(sourceAbsolutePath, destinationAbsolutePath))
```
Comment (johnterickson) [inline src/Common/Caching/CacheClient.cs:426]: Can we do this check at startup and warn if 1) repository is on volume with CoW *and* 2) nuget folder is not?

File: src/Common/Caching/CacheClient.cs (hunk 1 above)
Comment (erikmav) [reply src/Common/Caching/CacheClient.cs:470]: Complicated by `subst` as well as drive mounts and other obfuscations, see the complicated logic in the CoW library at https://github.com/microsoft/CopyOnWrite/blob/main/lib/Windows/VolumeEnumerator.cs . I would cut this method and the call to it and instead use `_copyOnWriteFilesystem.CopyOnWriteLinkSupportedBetweenPaths(sourceAbsolutePath, destinationAbsolutePath, CloneFlags.PathIsFullyResolved)` where it is called above. The logic in the CoW library has cached info and a fast lookup for each path to make it approximately as fast as checking drive letters in the 99% case.

File: src/Common/Caching/CacheClient.cs (hunk 3 above)
Comment (dfederm) [reply src/Common/Caching/CacheClient.cs]: Probably same with the cache actually

Comment (dfederm) [reply src/Common/Caching/CacheClient.cs]: image

File: src/Common/Fingerprinting/FingerprintFactory.cs [hunk 4]
```diff
@@ -211,8 +211,8 @@ private async Task SortAndAddInputFileHashesAsync(List<FingerprintEntry> entries
         foreach (string file in files)
         {
             string absoluteFilePath = pathsAreNormalized ? _pathNormalizer.Unnormalize(file) : file;
-            if (_pluginSettings.IgnoredInputPatterns.Count > 0
-                && !_pluginSettings.IgnoredInputPatterns.Any(pattern => pattern.IsMatch(absoluteFilePath)))
+            if (_pluginSettings.IgnoredInputPatterns.Count == 0
```
Comment (dfederm) [inline src/Common/Fingerprinting/FingerprintFactory.cs:214]: Unrelated fix for a regression caused by #58

**Review by johnterickson — APPROVED** [review]

File: src/Common/Caching/CacheClient.cs [hunk 5]
```diff
@@ -83,6 +88,8 @@ public abstract class CacheClient : ICacheClient
         {
             _outputHasher = new OutputHasher(_hasher);
         }
+
+        _canCloneInNugetCachePath = _copyOnWriteFilesystem.CopyOnWriteLinkSupportedInDirectoryTree(_nugetPackageRoot);
```
[the remark this answers is not in the document]
Comment (dfederm) [reply src/Common/Caching/CacheClient.cs:92]: We already check and warn in `WarnIfCowNotSupportedBetweenRepoRootAndPath`
